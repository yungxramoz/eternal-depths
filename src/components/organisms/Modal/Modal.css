.modal-container {
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  z-index: 100;
  height: 90dvh;
}

.modal-background {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5);
  display: flex;
  justify-content: center;
  align-items: center;
}

.modal {
  position: absolute;
  box-shadow: 0 0 10px 0 rgba(0, 0, 0, 0.5);
  min-width: 300px;
  max-width: 90vw;
  max-height: 90dvh;
  margin: 1em;
  z-index: 1;
}
