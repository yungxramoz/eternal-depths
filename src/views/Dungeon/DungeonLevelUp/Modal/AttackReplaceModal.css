.attack-list {
  display: flex;
  flex-direction: column;
  padding: 1em 0;
}
