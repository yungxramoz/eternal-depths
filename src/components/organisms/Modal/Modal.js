import React from 'react'
import ModalBody from '../../molecules/ModalBody/ModalBody'
import ModalFooter from '../../molecules/ModalFooter/ModalFooter'
import ModalHeader from '../../molecules/ModalHeader/ModalHeader'
import './Modal.css'
import RpgContainer from '../../templates/RpgContainer/RpgContainer'

const Modal = ({ title, children, onClose, dismissable = true }) => {
  if (onClose === undefined) {
    onClose = () => {}
  }

  return (
    <div className="modal-container">
      <RpgContainer golden className="modal">
        <ModalHeader title={title} />
        <ModalBody>{children}</ModalBody>
        <ModalFooter onClose={onClose} dismissable={dismissable} />
      </RpgContainer>
      <div className="modal-background" />
    </div>
  )
}

export default Modal
