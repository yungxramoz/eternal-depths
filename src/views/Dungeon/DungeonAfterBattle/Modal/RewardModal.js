import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import Modal from '../../../../components/organisms/Modal/Modal'
import RewardButton from '../../../../components/organisms/RewardButton/RewardButton'
import RpgSeparator from '../../../../components/templates/RpgSeperator/RpgSeparator'
import { RPGUI_ICON } from '../../../../constants/rpgui-icon'
import {
  equipItem,
  recoverHp,
} from '../../../../store/character/characterSlice'
import { calculatedHpReward } from '../../../../store/game/gameSlice'
import { generateItem } from '../../../../utils/item-generator'
import './RewardModal.css'

const RewardModal = ({ isOpen, setIsOpen }) => {
  const dispatch = useDispatch()

  const hpReward = useSelector(calculatedHpReward)
  const [gearReward] = useState(generateItem())

  const itemStats = () => {
    return Object.entries(gearReward.stats)
      .filter(([_, value]) => value > 0)
      .map(([key, value]) => `${key.slice(0, 3).toUpperCase()}:${value}`)
      .join(', ')
  }

  const selectGear = () => {
    dispatch(equipItem(gearReward))
    setIsOpen(false)
  }

  const selectHpRecovery = () => {
    dispatch(recoverHp(hpReward))
    setIsOpen(false)
  }
  return (
    <>
      {isOpen && (
        <Modal title="Encounter slayed!" dismissable={false}>
          <p>Choose your reward:</p>
          <div className="reward-selection-container">
            <RewardButton
              icon={gearReward.icon}
              title={gearReward.name}
              subtitle={gearReward.rarity}
              description={itemStats()}
              onClick={selectGear}
            ></RewardButton>
            <RpgSeparator golden />
            <RewardButton
              icon={RPGUI_ICON.POTION_RED}
              title="Food"
              description={`Recover ${hpReward} HP`}
              onClick={selectHpRecovery}
            ></RewardButton>
          </div>
        </Modal>
      )}
    </>
  )
}

export default RewardModal
