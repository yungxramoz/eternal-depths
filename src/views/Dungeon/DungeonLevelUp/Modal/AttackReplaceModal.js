import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import RpgButton from '../../../../components/atoms/RpgButton/RpgButton'
import Modal from '../../../../components/organisms/Modal/Modal'
import { characterReplaceAttack } from '../../../../store/game/gameSlice'
import ItemRadioButton from '../../../../components/organisms/ItemRadioButton/ItemRadioButton'

const AttackReplaceModal = ({ newAttack, onClose, onRewardSelected }) => {
  const dispatch = useDispatch()
  const attacks = useSelector((state) => state.game.character.current.attacks)
  const [selectedAttackId, setSelectedAttackId] = useState(2)
  const filteredAttacks = attacks.filter((attack) => attack.id !== 1)

  const handleReplace = () => {
    dispatch(characterReplaceAttack({ newAttack, attackId: selectedAttackId }))
    onRewardSelected()
  }

  return (
    <Modal
      title="Replace Attack"
      onDismiss={onClose}
      content={
        <div className="attack-replace-modal">
          <p>Select an attack to replace:</p>
          <form className="attack-list">
            {filteredAttacks.map((attack) => (
              <React.Fragment key={attack.id}>
                <ItemRadioButton
                  id={attack.id}
                  checked={selectedAttackId === attack.id}
                  title={attack.name}
                  subtitle={`Cooldown: ${attack.cooldown}`}
                  description={attack.description}
                  onChange={() => setSelectedAttackId(attack.id)}
                />
              </React.Fragment>
            ))}
          </form>
        </div>
      }
      footerContent={
        <RpgButton
          text="Replace"
          disabled={selectedAttackId == null}
          onClick={handleReplace}
        />
      }
    />
  )
}

export default AttackReplaceModal
