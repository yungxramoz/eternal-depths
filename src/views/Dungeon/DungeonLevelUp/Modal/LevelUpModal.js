import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import Modal from '../../../../components/organisms/Modal/Modal'
import RewardButton from '../../../../components/organisms/RewardButton/RewardButton'
import RpgSeparator from '../../../../components/templates/RpgSeperator/RpgSeparator'
import { RPGUI_ICON } from '../../../../constants/rpgui-icon'
import './LevelUpModal.css'
import { generateAttack } from '../../../../utils/attack'
import { characterLearnAttack } from '../../../../store/game/gameSlice'

const LevelUpModal = ({ isOpen, setIsOpen }) => {
  const dispatch = useDispatch()
  const characterLevel = useSelector(
    (state) => state.game.character.current.level,
  )
  const characterAttacks = useSelector(
    (state) => state.game.character.current.attacks,
  )
  const [attackReward] = useState(generateAttack())

  const selectAttackReward = () => {
    dispatch(characterLearnAttack(attackReward))
    setIsOpen(false)
  }

  const selectStatReward = () => {
    console.log('Stat reward selected')
    setIsOpen(false)
  }

  return (
    <>
      {isOpen && (
        <Modal title={`Level ${characterLevel} reached!`} dismissable={false}>
          <p>Choose your reward:</p>
          <div className="reward-selection-container">
            <RewardButton
              icon={RPGUI_ICON.BOOK}
              title={attackReward.name}
              subtitle={`Cooldown: ${attackReward.cooldown}`}
              description={attackReward.description}
              disabled={characterAttacks.length >= 3}
              onClick={selectAttackReward}
            ></RewardButton>
            <RpgSeparator golden />
            <RewardButton
              icon={RPGUI_ICON.POTION}
              title="Attribute"
              description="Improve 2 attributes"
              onClick={selectStatReward}
            ></RewardButton>
          </div>
        </Modal>
      )}
    </>
  )
}

export default LevelUpModal
