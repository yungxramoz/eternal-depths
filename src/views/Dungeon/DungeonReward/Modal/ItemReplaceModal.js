import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import RpgButton from '../../../../components/atoms/RpgButton/RpgButton'
import Modal from '../../../../components/organisms/Modal/Modal'
import { characterEquipItem } from '../../../../store/game/gameSlice'
import ItemDetails from '../../../../components/molecules/ItemDetails/ItemDetails'
import RpgSeparator from '../../../../components/templates/RpgSeperator/RpgSeparator'
import { getItemStats } from '../../../../utils/item-generator'

const ItemReplaceModal = ({ newItem, onClose, onRewardSelected }) => {
  const dispatch = useDispatch()
  const items = useSelector((state) => state.game.character.current.items)
  const currentItem = items[newItem.type]

  const itemStats = (item) => getItemStats(item)

  const handleReplace = () => {
    dispatch(characterEquipItem(newItem))
    onRewardSelected()
  }

  return (
    <Modal
      title="Replace Item"
      onDismiss={onClose}
      content={
        <div className="item-replace-modal">
          <p>Do you want to replace the current item:</p>
          <ItemDetails
            title={currentItem.name}
            subtitle={currentItem.rarity}
            description={itemStats(currentItem)}
            icon={currentItem.icon}
          />
          <RpgSeparator />
          <p>New item:</p>
          <ItemDetails
            title={newItem.name}
            subtitle={newItem.rarity}
            description={itemStats(newItem)}
            icon={newItem.icon}
          />
        </div>
      }
      footerContent={<RpgButton text="Replace" onClick={handleReplace} />}
    />
  )
}

export default ItemReplaceModal
