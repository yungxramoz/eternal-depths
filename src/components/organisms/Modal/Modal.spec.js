import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import Modal from './Modal'

describe('Modal', () => {
  it('renders correctly', () => {
    render(
      <Modal title="Test Title" onClose={() => {}}>
        <div>Test Content</div>
      </Modal>,
    )
    expect(document.body).toMatchSnapshot('Modal')
  })

  it('calls onClose prop when "Close" button is clicked', () => {
    const onClose = jest.fn()
    render(<Modal title="Test Title" onClose={onClose} />)
    fireEvent.click(screen.getByText('Close'))
    expect(onClose).toHaveBeenCalledTimes(1)
  })

  it('does not render "Close" button when dismissable prop is false', () => {
    render(<Modal title="Test Title" onClose={() => {}} dismissable={false} />)
    expect(screen.queryByText('Close')).not.toBeInTheDocument()
  })
})
