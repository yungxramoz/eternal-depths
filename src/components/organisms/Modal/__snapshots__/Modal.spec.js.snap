// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders correctly: Modal 1`] = `
<body>
  <div>
    <div
      class="modal-container"
    >
      <div
        class="rpgui-container container framed framed-golden modal"
      >
        <div
          class="modal-header"
        >
          <h2>
            Test Title
          </h2>
          <hr
            class="separator"
          />
        </div>
        <div
          class="modal-body"
        >
          <div>
            Test Content
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <hr
            class="separator"
          />
          <button
            class="rpgui-button btn"
          >
            Close
          </button>
        </div>
      </div>
      <div
        class="modal-background"
      />
    </div>
  </div>
</body>
`;
